modules/core/src/test/java/org/apache/ignite/internal/processors/continuous/ContinuousQueryBuffersCleanupTest.java

```diff
@@ -0,0 +1,238 @@
+                checkBuffersCleared(primaryNodeIdx, qry -> cli.cache(DEFAULT_CACHE_NAME).query(qry));
+                checkBuffersCleared(primaryNodeIdx, qry -> thinCli.cache(DEFAULT_CACHE_NAME).query(qry));
+            }
+        }
+    }
+
+    /** */
+    private void checkBuffersCleared(int primaryNodeIdx, Function<Query<?>, QueryCursor<?>> queryExecutor) throws Exception {
+        IgniteEx primary = grid(primaryNodeIdx);
+
+        int cacheOpRounds = Math.round(DFLT_CONTINUOUS_QUERY_BACKUP_ACK_THRESHOLD * 0.3f);
+
+        // We repeatedly perform 5 cache operations that raise CREATE, UPDATE, REMOVED, EXPIRIED events.
+        // The total number of events is selected in a such way as to check for a backup notification due to a
+        // buffer overflow, and then by timeout.
+        int expEventsCnt = cacheOpRounds * 5;
+
+        CountDownLatch cqListenerNotifiedLatch = new CountDownLatch(expEventsCnt);
+
+        Query<?> qry = new ContinuousQuery<>()
+            .setLocalListener(events -> cqListenerNotifiedLatch.countDown())
+            .setPageSize(1)
+            .setIncludeExpired(true);
+
+        try (QueryCursor<?> ignored = queryExecutor.apply(qry)) {
+            for (int i = 0; i < cacheOpRounds; i++) {
+                int testKey = keyForNode(primary);
+
+                primary.cache(DEFAULT_CACHE_NAME).put(testKey, 0);
+                primary.cache(DEFAULT_CACHE_NAME).put(testKey, 1);
+                primary.cache(DEFAULT_CACHE_NAME).remove(testKey);
+                primary.cache(DEFAULT_CACHE_NAME)
+                    .withExpiryPolicy(new CreatedExpiryPolicy(new Duration(TimeUnit.MILLISECONDS, 100)))
+                    .put(keyForNode(primary), 0);
+            }
+
+            assertTrue(cqListenerNotifiedLatch.await(getTestTimeout(), MILLISECONDS));
+
+            assertTrue(waitForCondition(() -> G.allGrids().stream().allMatch(this::isContinuesQueryBufferEmtpy), getTestTimeout()));
+        }
+    }
+
+    /** */
+    private boolean isContinuesQueryBufferEmtpy(Ignite ignite) {
+        GridContinuousProcessor contProc = ((IgniteEx)ignite).context().continuous();
+
+        Collection<CacheContinuousQueryHandler<?, ?>> cqHandlers = new ArrayList<>();
+
+        contProc.remoteRoutineInfos().forEach((routineId, rmtRoutineInfo) ->
+            cqHandlers.add((CacheContinuousQueryHandler<?, ?>)rmtRoutineInfo.handler()));
+
+        contProc.localRoutineInfos().forEach((routineId, locRoutineInfo) ->
+            cqHandlers.add((CacheContinuousQueryHandler<?, ?>)locRoutineInfo.handler()));
+
+        contProc.clientRoutineInfos().forEach((nodeId, rmtRoutineInfos) ->
+            rmtRoutineInfos.forEach((routineId, locRoutineInfo) ->
+                cqHandlers.add((CacheContinuousQueryHandler<?, ?>)locRoutineInfo.handler())));
+
+        for (CacheContinuousQueryHandler<?, ?> cqHnd : cqHandlers) {
+            ConcurrentMap<Integer, CacheContinuousQueryEventBuffer> entryBufs = U.field(cqHnd, "entryBufs");
```

Comment:
I don't like that a lot of reflection used by this test. In the same time you rename some variable and use it name in reflection, it's fragile logic. I know that there are two different packages for continuous queries and you need to access internal class fields in both packages and can't use package private methods. But, perhaps, we can introduce some class like `ContinuousQueriesTestUtils` in the package `org.apache.ignite.internal.processors.cache.query.continuous` end expose some methods to use in tests. WDYT?

Reply:
Fixed.